common/dal2/src/services/s3.rs

```diff
@@ -99,20 +125,60 @@ impl Builder {
 
         let mut cfg = AwsS3::config::Builder::from(&aws_cfg);
 
-        // Load users input first, if user not input, we will fallback to aws
-        // default load logic.
-        if let Some(endpoint) = self.endpoint {
-            cfg = cfg.endpoint_resolver(AwsS3::Endpoint::immutable(
-                http::Uri::from_str(&endpoint).map_err(|_| Error::BackendConfigurationInvalid {
-                    key: "endpoint".to_string(),
-                    value: endpoint.to_string(),
-                })?,
-            ));
+        // ## Safety
+        //
+        // aws s3 sdk requires a static str for region, so we crate a static string here.
+        // This value is scoped under `finish` function only, changed once and used
+        // once while initializing the client, so it must be safe.
+        static mut REGION: String = String::new();
+        unsafe { REGION = self.region.clone() };
+        // TODO: Maybe we can
+        //
+        // - use "default" as the default region.
+        // - use "us-east-1" as the default region.
+        // - detect the region at runtime via `ListBuckets`.
+        cfg = cfg.region(AwsS3::Region::new(unsafe { REGION.as_str() }));
```

Comment:
Here is a bit strange, looks like we should use some AwsS3 function to do a convert from self.region:
https://docs.rs/aws-region/0.23.3/src/awsregion/region.rs.html#137

Reply:
Oh, makes sense. I will upate it.

Reply:
I give `aws-region` a try, it's not a part of aws rust sdk, and we can't convert by this crate.

I will figure a better to handle the endpoint and region.

Reply:
`impl<'a> From<String> for Cow<'a, str>`

https://doc.rust-lang.org/std/borrow/enum.Cow.html#impl-From%3CString%3E

Reply:
It works! Thanks a lot!

Reply:
Well, I still have question on this.

`Cow` works well for `Cow<'static, str>`, but aws `Metadata` requires a `&static str`.

```rust
#[derive(Debug)]
pub struct Metadata {
    /// URI for the endpoint.
    ///
    /// May contain `{region}` which will replaced with the region during endpoint construction
    pub uri_template: &'static str,

    /// Protocol to use for this endpoint
    pub protocol: Protocol,

    /// Credential scope to set for requests to this endpoint
    pub credential_scope: CredentialScope,

    /// Signature versions supported by this endpoint.
    ///
    /// Currently unused since the SDK only supports SigV4
    pub signature_versions: SignatureVersion,
}
```

Maybe it's better to implement a `ResolveAwsEndpoint` byself?

Reply:
I will refactor region and endpoint parse logic.